components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/XACMLScopeValidator.java

```diff
@@ -0,0 +1,244 @@
+import org.wso2.carbon.identity.application.authentication.framework.util.FrameworkUtils;
+import org.wso2.carbon.identity.entitlement.EntitlementException;
+import org.wso2.carbon.identity.entitlement.PDPConstants;
+import org.wso2.carbon.identity.entitlement.common.EntitlementPolicyConstants;
+import org.wso2.carbon.identity.entitlement.common.dto.RequestDTO;
+import org.wso2.carbon.identity.entitlement.common.dto.RowDTO;
+import org.wso2.carbon.identity.entitlement.common.util.PolicyCreatorUtil;
+import org.wso2.carbon.identity.oauth.common.exception.InvalidOAuthClientException;
+import org.wso2.carbon.identity.oauth.dao.OAuthAppDO;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.internal.OAuth2ServiceComponentHolder;
+import org.wso2.carbon.identity.oauth2.model.AccessTokenDO;
+import org.wso2.carbon.identity.oauth2.util.OAuth2Util;
+
+import java.io.ByteArrayInputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import javax.xml.stream.XMLStreamException;
+
+
+/**
+ * The Scope Validation implementation. This uses XACML policies to evaluate scope validation defined by the user.
+ */
+public class XACMLScopeValidator extends OAuth2ScopeValidator {
+
+    private static final String DECISION_XPATH = "//ns:Result/ns:Decision/text()";
+    private static final String XACML_NS_PREFIX = "ns";
+    private static final String ACTION_VALIDATE = "token_validation";
+    private static final String RULE_EFFECT_PERMIT = "Permit";
+    private static final String RULE_EFFECT_NOT_APPLICABLE = "NotApplicable";
+
+    private static final String ACTION_CATEGORY = "http://wso2.org/identity/identity-action";
+    private static final String SP_CATEGORY = "http://wso2.org/identity/sp";
+    private static final String USER_CATEGORY = "http://wso2.org/identity/user";
+    private static final String SCOPE_CATEGORY = "http://wso2.org/identity/oauth-scope";
+
+    private static final String AUTH_ACTION_ID = ACTION_CATEGORY + "/action-name";
+    private static final String SP_NAME_ID = SP_CATEGORY + "/sp-name";
+    private static final String USERNAME_ID = USER_CATEGORY + "/username";
+    private static final String USER_STORE_ID = USER_CATEGORY + "/user-store-domain";
+    private static final String USER_TENANT_DOMAIN_ID = USER_CATEGORY + "/user-tenant-domain";
+    private static final String SCOPE_ID = SCOPE_CATEGORY + "/scope-name";
+
+    private Log log = LogFactory.getLog(XACMLScopeValidator.class);
+
+    @Override
+    public boolean validateScope(AccessTokenDO accessTokenDO, String resource) throws IdentityOAuth2Exception {
+
+        if (isUnAuthorizedToken(accessTokenDO)) {
+            return false;
+        }
+        String authzUser = accessTokenDO.getAuthzUser().getUserName();
+        boolean isValidated = false;
+        FrameworkUtils.startTenantFlow(accessTokenDO.getAuthzUser().getTenantDomain());
```

Comment:
don't we need to start the tenant flow inside try block?

Reply:
fixed with https://github.com/wso2-extensions/identity-oauth-addons/pull/22/commits/687f6b819361687150ed88fff5adc7d16f56d457